File: packages/go/internal/seguranca/ataques/ataque_independencia_proposito.go
package ataques

import (
	"encoding/hex"
	"fmt"

	"criptografia/internal/seguranca"
)

// OpcoesIndependenciaProposito configura o AtaqueIndependenciaProposito.
type OpcoesIndependenciaProposito struct {
	Amostras int
	Tamanho  int
}

// AtaqueIndependenciaProposito verifica se os keystreams de propósitos
// diferentes ('enc' e 'mac') derivados da mesma (key, iv) são independentes.
// Se não forem, recuperar o keystream de dados permite derivar a chave do MAC
// e forjar tokens válidos.
type AtaqueIndependenciaProposito struct {
	amostras int
	tamanho  int
}

// NovoAtaqueIndependenciaProposito cria o ataque. Sem argumento, usa
// amostras = 2000 e tamanho = 32.
func NovoAtaqueIndependenciaProposito(opcoes ...OpcoesIndependenciaProposito) *AtaqueIndependenciaProposito {
	o := OpcoesIndependenciaProposito{Amostras: 2000, Tamanho: 32}
	if len(opcoes) > 0 {
		o = opcoes[0]
	}
	return &AtaqueIndependenciaProposito{amostras: o.Amostras, tamanho: o.Tamanho}
}

// Nome devolve o nome exibido no relatório.
func (a *AtaqueIndependenciaProposito) Nome() string {
	return "Independência entre keystreams de propósitos diferentes (enc x mac)"
}

// Executar roda o ataque.
func (a *AtaqueIndependenciaProposito) Executar(alvo seguranca.AlvoCriptografico) (seguranca.ResultadoAtaque, error) {
	tamChave := len(alvo.ChaveDeTeste())

	distancias := make([]float64, 0, a.amostras)
	mascaras := map[string]bool{}

	for i := 0; i < a.amostras; i++ {
		chave := seguranca.RandomBytes(tamChave)
		iv := seguranca.RandomBytes(alvo.TamanhoIv())

		enc := alvo.GerarKeystreamBruto(chave, iv, "enc", a.tamanho)
		mac := alvo.GerarKeystreamBruto(chave, iv, "mac", a.tamanho)

		distancias = append(distancias, float64(seguranca.BitsDiferentes(enc, mac))/float64(a.tamanho*8))
		mascaras[hex.EncodeToString(a.xorBytes(enc, mac))] = true
	}

	media := 0.0
	for _, v := range distancias {
		media += v
	}
	media /= float64(len(distancias))
	mascarasDistintas := len(mascaras)

	vulneravel := media < 0.4 || media > 0.6 || mascarasDistintas < a.amostras
	severidade := seguranca.SevInfo
	if vulneravel {
		severidade = seguranca.SevCritica
	}

	sufixo := ""
	if mascarasDistintas < a.amostras {
		sufixo = " - MÁSCARA REPETIDA: mac previsível a partir de enc!"
	}

	return seguranca.ResultadoAtaque{
		NomeAtaque: a.Nome(),
		Vulneravel: vulneravel,
		Severidade: severidade,
		Detalhes: fmt.Sprintf(
			"Hamming médio enc x mac=%.1f%% (esperado ~50%%); %d máscara(s) XOR distinta(s) em %d amostras%s",
			media*100, mascarasDistintas, a.amostras, sufixo,
		),
		Dados: map[string]any{"media": media, "mascaras_distintas": mascarasDistintas},
	}, nil
}

// xorBytes faz XOR byte a byte de dois buffers.
func (a *AtaqueIndependenciaProposito) xorBytes(a2, b []byte) []byte {
	n := min(len(a2), len(b))
	out := make([]byte, n)
	for i := 0; i < n; i++ {
		out[i] = a2[i] ^ b[i]
	}
	return out
}

File: packages/java/src/main/java/fbc/Seguranca/Ataques/AtaqueIndependenciaProposito.java
package fbc.Seguranca.Ataques;

import fbc.Seguranca.AlvoCriptografico;
import fbc.Seguranca.AtaqueInterface;
import fbc.Seguranca.ResultadoAtaque;
import fbc.Seguranca.Severidade;
import fbc.Seguranca.SkipAtaqueException;
import fbc.Seguranca.Util;
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Locale;
import java.util.Map;
import java.util.Set;

/**
 * A mesma (key, iv) é usada pra derivar o keystream de dados ('enc') e a
 * chave do MAC ('mac'). Se os dois streams não forem independentes, um
 * atacante que recupere o keystream de dados (ataque de texto conhecido)
 * pode derivar a chave do MAC e FORJAR tokens válidos.
 *
 * O teste procura dois sintomas de acoplamento:
 *  1. a distância de Hamming média entre os streams deve ser ~50%;
 *  2. XOR(enc, mac) NÃO pode se repetir entre (key, iv) diferentes - se for
 *     uma máscara fixa, o mac é 100% previsível a partir do enc.
 */
public class AtaqueIndependenciaProposito implements AtaqueInterface {
    private final int amostras;
    private final int tamanho;

    public AtaqueIndependenciaProposito() {
        this(2000, 32);
    }

    public AtaqueIndependenciaProposito(int amostras, int tamanho) {
        this.amostras = amostras;
        this.tamanho = tamanho;
    }

    @Override
    public String nome() {
        return "Independência entre keystreams de propósitos diferentes (enc x mac)";
    }

    @Override
    public ResultadoAtaque executar(AlvoCriptografico alvo) {
        byte[] primeiro = alvo.gerarKeystreamBruto(
                alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8),
                Util.randomBytes(alvo.tamanhoIv()),
                "enc",
                this.tamanho);
        if (primeiro == null) {
            throw new SkipAtaqueException("Alvo não expõe gerarKeystreamBruto.");
        }

        int lenChave = alvo.chaveDeTeste().getBytes(StandardCharsets.UTF_8).length;
        double soma = 0.0;
        Set<String> mascaras = new HashSet<>();
        for (int i = 0; i < this.amostras; i++) {
            byte[] chave = Util.randomBytes(lenChave);
            byte[] iv = Util.randomBytes(alvo.tamanhoIv());

            byte[] enc = alvo.gerarKeystreamBruto(chave, iv, "enc", this.tamanho);
            byte[] mac = alvo.gerarKeystreamBruto(chave, iv, "mac", this.tamanho);

            soma += (double) Util.bitsDiferentes(enc, mac) / (this.tamanho * 8);
            mascaras.add(Util.bytesParaHex(this.xorBytes(enc, mac)));
        }

        double media = soma / this.amostras;
        int mascarasDistintas = mascaras.size();
        boolean vulneravel = media < 0.4 || media > 0.6 || mascarasDistintas < this.amostras;

        Map<String, Object> dados = new HashMap<>();
        dados.put("media", media);
        dados.put("mascaras_distintas", mascarasDistintas);

        return new ResultadoAtaque(
                nome(),
                vulneravel,
                vulneravel ? Severidade.CRITICA : Severidade.INFO,
                String.format(
                        Locale.ROOT,
                        "Hamming médio enc x mac=%.1f%% (esperado ~50%%); %d máscara(s) XOR distinta(s) em %d amostras%s",
                        media * 100,
                        mascarasDistintas,
                        this.amostras,
                        mascarasDistintas < this.amostras
                                ? " - MÁSCARA REPETIDA: mac previsível a partir de enc!"
                                : ""),
                dados);
    }

    private byte[] xorBytes(byte[] a, byte[] b) {
        int len = Math.min(a.length, b.length);
        byte[] out = new byte[len];
        for (int i = 0; i < len; i++) {
            out[i] = (byte) ((a[i] ^ b[i]) & 0xFF);
        }
        return out;
    }
}
